### PR TITLE
Add base Scorer

### LINKED ISSUE #2: Implement Scorer Base Implementation
```python

s = Scorer(stuff)

m = Mapper.suggest_mapping(m1, m2)

f : float = s.score(m)
```

### PR DESCRIPTION
**Pull request by dwhswenson**
Resolves #2.

## Overview

This provides a base class `Scorer`. The `Scorer.__call__` returns a `NamedTuple` with attributes `score` and `annotation`. Subclass implementers **must** implement the method `score`, and **may** override the method `annotation`.

## Self-criticism

* The module-level docstring is pretty thin. I just wasn't sure what else to say at this point.
* There's perhaps a little too much detail on implementation in the `Scorer` class docstring. I'm concerned about this getting out-of-sync with code, but it will be important to document the idea. Any better approach welcome.
* This is heavier in precise use of typing than I usually care about, but for abstract classes, I think detalied typing might be worth it.
* There's an overly-trivial placeholder fixture for a `mock_atommapping`. Since we don't have a completed `AtomMapping` class, we can't create a good fixture for it yet, but I added the stub.

### CODE REVIEW

**Review by richardjgowers — APPROVED** [review]
Looks good.  I did have somewhere in the back of my mind that only requiring `__call__` would allow functions to also be used, obviously methods break that but maybe that's not such a drama.

**Comment by dwhswenson** [comment]
> Looks good. I did have somewhere in the back of my mind that only requiring __call__ would allow functions to also be used, obviously methods break that but maybe that's not such a drama.

Implementers could still write a function, as long as it returns a (duck type of a) `ScoreAnnotation`. Maybe it would be best to make `score` and `annotation` private methods if that case is likely to be used, so they aren't used independently of `__call__`?

File: openfe/setup/scorer.py [hunk 1]
```diff
@@ -1,4 +1,71 @@
...
+        ----------
+        atommapping : AtomMapping
+            AtomMapping to score
+
+        Returns
+        -------
+        Union[float, None] :
+            The score, or ``None`` if no score is calculated
+        """
+        raise NotImplementedError()
```
Comment (IAlibay) [inline openfe/setup/scorer.py:52]: I know this _shouldn't_ be ever used on it's own, but for the sake of someone doing something silly + coverage can we a) get a test that triggers this, b) a simple message to tell folks to not call the base class directly?

Comment (dwhswenson) [reply openfe/setup/scorer.py:52]: I usually use the rule that `raise NotImplementedError` is part of the coverage exclude lines (along with `__repr__`); do we want to make a practice of always testing `NotImplementedError`s? (I'm fine either way.) Will definitely add a clearer message, though.

Comment (IAlibay) [reply openfe/setup/scorer.py:52]: I'm not super fussed either way, I do find trying to trigger errors cathartic so I'm happy just adding them randomly on a bored weekend evening.

Comment (dwhswenson) [reply openfe/setup/scorer.py]: I'll went ahead and added the test -- it isn't hard and doesn't hurt anything. We can always ignore in coverage later (having a test that covers lines that don't count toward coverage isn't a problem).

Speaking of coverage, CodeCov is being flaky. I'm trying to dig up some old CodeCov problems I've had (and their solutions) to see if it is something I've seen before.

**Comment by dwhswenson** [comment]
Shouldn't pep8speaks have had something to say about this PR? Is it not enabled for PRs coming from forks?

**Comment by IAlibay** [comment]
> Shouldn't pep8speaks have had something to say about this PR? Is it not enabled for PRs coming from forks?

Yeah I'm not sure why it's not being triggered... default options _should_ work on forks.

**Review by IAlibay — APPROVED** [review]

**Pull request merged**